README.md

```diff
@@ -2,11 +2,876 @@
+
+```java
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.Tracer;
+
+@RestController
+@RequestMapping(value = "/time")
+public class SecondServiceController {
+  @Autowired
+  private Tracer tracer;
+
+  @GetMapping
+  public String secondTracedMethod() {
+    return "It's time to get a watch";
+  }
+}
+```
+
+#### Create Controller Interceptor
+
+Add the class below to wrap all requests to the SecondServiceController in a span. This class will call the preHandle method before the rest controller is entered and the postHandle method after a response is created. 
+
+The preHandle method starts a span for each request. The postHandle method closes the span and adds the span context to the response header. This implementation is shown below:   
+
+```java
+import java.util.logging.Logger;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.web.servlet.HandlerInterceptor;
+import org.springframework.web.servlet.ModelAndView;
+import io.grpc.Context;
+import io.opentelemetry.OpenTelemetry;
+import io.opentelemetry.context.propagation.HttpTextFormat;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.Tracer;
+import io.opentelemetry.trace.TracingContextUtils;
+
+@Component
+public class ControllerTraceInterceptor implements HandlerInterceptor {
+
+  private static HttpTextFormat.Getter<HttpServletRequest> getter =
+      new HttpTextFormat.Getter<HttpServletRequest>() {
+        public String get(HttpServletRequest req, String key) {
+          return req.getHeader(key);
+        }
+      };
+
+  private static HttpTextFormat.Setter<HttpServletResponse> setter =
+      new HttpTextFormat.Setter<HttpServletResponse>() {
+        @Override
+        public void set(HttpServletResponse response, String key, String value) {
+          response.addHeader(key, value);
+        }
+      };
+
+  @Autowired
+  private Tracer tracer;
+
+  @Override
+  public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
+      throws Exception {
+    Span span;
+    try {
+
+      Context context = OpenTelemetry.getPropagators().getHttpTextFormat()
+          .extract(Context.current(), request, getter);
+
+      span = tracer.spanBuilder(request.getRequestURI())
+          .setParent(TracingContextUtils.getSpan(context)).startSpan();
+      span.setAttribute("handler", "pre");
+    } catch (Exception e) {
+      span = tracer.spanBuilder(request.getRequestURI()).startSpan();
```

Comment:
This looks a bit strange - why do you have to open the span twice? Why not open it once, and put diferent attributes? The duplication here is unclear.

Reply:
Here I try to start a new span and set the parent span (if it exists). If there is no parent span (TracingContextUtils.getSpan(context) returns an invalid span) this line throws an exception. In the catch block we create a new span with out a parent. This is how all the otel tutorials I came across approached span propagation. Instead of this approach I could do a check. Now that you mention it, it seems like the better approach:
   ```
    private Span setParentSpan(HttpServletRequest request, Context context) {
    Span parentSpan = TracingContextUtils.getSpan(context);
    if (parentSpan.getContext().isValid()) {
      return tracer.spanBuilder(request.getRequestURI()).setParent(parentSpan).startSpan();
    }

    Span span = tracer.spanBuilder(request.getRequestURI()).startSpan();
    span.addEvent("Parent Span Not Found");

    return 
  ```

Reply:
Ack. Even if you did use a try/catch, try to avoid as much as possible any code replication. It's often possible to structure your setup so that it can be reused in both cases.